pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java

```diff
@@ -243,75 +228,35 @@ private Map<String, Map<String, String>> reassignSegments(String instancePartiti
-        newAssignment = SegmentAssignmentUtils
-            .rebalanceTableWithHelixAutoRebalanceStrategy(currentAssignment, instances, _replication);
-      } else {
-        // Replica-group based assignment
-
-        if (_partitionColumn == null) {
-          // NOTE: Shuffle the segments within the current assignment to avoid moving only new segments to the new added
-          //       servers, which might cause hotspot servers because queries tend to hit the new segments. Use the
-          //       table name hash as the random seed for the shuffle so that the result is deterministic.
-          List<String> segments = new ArrayList<>(currentAssignment.keySet());
-          Collections.shuffle(segments, new Random(_offlineTableName.hashCode()));
-
-          newAssignment = new TreeMap<>();
-          SegmentAssignmentUtils
-              .rebalanceReplicaGroupBasedPartition(currentAssignment, instancePartitions, 0, segments, newAssignment);
-        } else {
-          newAssignment = rebalanceTableWithPartition(currentAssignment, instancePartitions);
-        }
-      }
+      newAssignment = _segmentAssignmentStrategy
+          .reassignSegments(currentAssignment, instancePartitions, InstancePartitionsType.OFFLINE);
     }
     return newAssignment;
   }
 
-  private Map<String, Map<String, String>> rebalanceTableWithPartition(
-      Map<String, Map<String, String>> currentAssignment, InstancePartitions instancePartitions) {
-    // Fetch partition id from segment ZK metadata
-    List<SegmentZKMetadata> segmentsZKMetadata =
-        ZKMetadataProvider.getSegmentsZKMetadata(_helixManager.getHelixPropertyStore(), _offlineTableName);
-    Map<String, SegmentZKMetadata> segmentZKMetadataMap = new HashMap<>();
-    for (SegmentZKMetadata segmentZKMetadata : segmentsZKMetadata) {
-      segmentZKMetadataMap.put(segmentZKMetadata.getSegmentName(), segmentZKMetadata);
-    }
-    int numPartitions = instancePartitions.getNumPartitions();
-    Map<Integer, List<String>> instancePartitionIdToSegmentsMap = new HashMap<>();
-    for (String segmentName : currentAssignment.keySet()) {
-      int partitionId = getPartitionId(segmentZKMetadataMap.get(segmentName));
-      int instancePartitionId = partitionId % numPartitions;
-      instancePartitionIdToSegmentsMap.computeIfAbsent(instancePartitionId, k -> new ArrayList<>()).add(segmentName);
-    }
-
-    // NOTE: Shuffle the segments within the current assignment to avoid moving only new segments to the new added
-    //       servers, which might cause hotspot servers because queries tend to hit the new segments. Use the table
-    //       name hash as the random seed for the shuffle so that the result is deterministic.
-    Random random = new Random(_offlineTableName.hashCode());
-    for (List<String> segments : instancePartitionIdToSegmentsMap.values()) {
-      Collections.shuffle(segments, random);
+  private void getSegmentAssignmentStrategy(
```

Comment:
We can take this function out to the common class `e.g. SegmentAssignmentStrategyFactory`

```
Map<InstancePartitionsType, SegmentAssignmentStrategy> getSegmentAssignmentStrategy(inputs...)
```

and call this during the init and only cache the assignment strategy from `OFFLINE`

Reply:
updated!